tsdb/chunks/chunks.go

```diff
@@ -284,18 +299,44 @@ func MergeChunks(a, b chunkenc.Chunk) (*chunkenc.XORChunk, error) {
 	return newChunk, nil
 }
 
+// WriteChunks writes as many chunks as possible to the current segment,
+// cuts a new segment when the current segment is full and
+// writes the rest of the chunks in the new segment.
 func (w *Writer) WriteChunks(chks ...Meta) error {
-	// Calculate maximum space we need and cut a new segment in case
-	// we don't fit into the current one.
-	maxLen := int64(binary.MaxVarintLen32) // The number of chunks.
-	for _, c := range chks {
-		maxLen += binary.MaxVarintLen32 + 1 // The number of bytes in the chunk and its encoding.
-		maxLen += int64(len(c.Chunk.Bytes()))
-		maxLen += 4 // The 4 bytes of crc32
+	var (
+		chksBatchSize int64
+		end           int
+	)
+
+	for _, chk := range chks {
+		// Each chunk contains: data length + encoding + the data itself + crc32
+		chksBatchSize += int64(maxChunkLengthFieldSize) // The data length is a variable length field so use the maximum possible value.
+		chksBatchSize += chunkEncodingSize              // The chunk encoding.
+		chksBatchSize += int64(len(chk.Chunk.Bytes()))  // The data itself.
+		chksBatchSize += crc32.Size                     // The 4 bytes of crc32
+
+		end++
+		if chksBatchSize+w.n > w.segmentSize {
+			if err := w.writeChunks(chks[:end], true); err != nil {
+				return err
+			}
+			chks = chks[end:]
+			chksBatchSize = 0
+			end = 0
+			continue
+		}
 	}
-	newsz := w.n + maxLen
 
-	if w.wbuf == nil || newsz > w.segmentSize && maxLen <= w.segmentSize {
+	return w.writeChunks(chks, false)
+}
+
+func (w *Writer) writeChunks(chks []Meta, cutChunk bool) error {
+	if len(chks) == 0 {
+		return nil
+	}
+	// w.wbuf == nil means it is the first chunk
+	// so need to start a new segment.
+	if cutChunk || w.wbuf == nil {
```

Comment:
`chks` should be written to the current segment.

if `cutChunk` is true, cut after writing `chks`?

Reply:
Yes that will be done, see https://github.com/krasi-georgiev/prometheus/pull/4